### PR TITLE
Export classifications in segments

### PR DESCRIPTION
**Pull request by marten**
Describe your change here.

# Review checklist

- [ ] First, the most important one: is this PR small enough that you can actually review it? Feel free to just reject a branch if the changes are hard to review due to the length of the diff.
- [ ] If there are any migrations, will they the previous version of the app work correctly after they've been run (e.g. the don't remove columns still known about by ActiveRecord).
- [ ] If anything changed with regards to the public API, are those changes also documented in the `apiary.apib` file?
- [ ] Are all the changes covered by tests? Think about any possible edge cases that might be left untested.

### CODE REVIEW

File: db/migrate/20170509105754_create_classifications_export_segments.rb [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+      t.integer :last_classification_id, null: false
+      t.integer :requester_id, null: false
+
+      t.datetime :started_at
+      t.datetime :finished_at
+
+      t.timestamps null: false
+    end
+
+    add_foreign_key :classifications_export_segments, :classifications, column: :first_classification_id
```
Comment (marten) [inline db/migrate/20170509105754_create_classifications_export_segments.rb]: Add `NOT VALID` as per https://joinhandshake.com/engineering/2017/02/27/database-migrations-and-locks.html

File: app/operations/classifications_export_segments/create.rb [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        segment
+      else
+        nil
+      end
+    end
+
+    private
+
+    def workflow
+      @workflow ||= Workflow.find(links[:workflow])
```
Comment (marten) [inline app/operations/classifications_export_segments/create.rb]: This doesn't do any permissions checking!

File: spec/controllers/api/v1/classifications_export_segments_controller_spec.rb [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    let(:index_params) { {project_id: project.id} }
+    let(:n_visible) { 1 }
+
+    it_behaves_like "is indexable", false
+
+    it 'should not allow access to projects the user is not a collaborator on' do
+      project2 = create :project
+      segment2 = create :classifications_export_segment, project: project
+      get :index, project_id: project2.id
+      expect(response.status).to eq(401)
```
Comment (camallen) [inline spec/controllers/api/v1/classifications_export_segments_controller_spec.rb]: Is this 401 coming from the create operation?

File: spec/factories/classifications.rb [hunk 4]
```diff
@@ -14,8 +14,9 @@
     workflow_version "15.15"
     completed true
     user
-    project
     workflow
+    project { workflow&.project || build(:project) }
```
Comment (camallen) [inline spec/factories/classifications.rb]: nice

File: app/models/classifications_export_segment.rb [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    segment = self.class.new(project_id: project_id, workflow_id: workflow_id, requester: requester)
+    segment.set_first_last_classifications(last_classification_id)
+    segment
+  end
+
+  def set_first_last_classifications(after)
+    scope = complete_classifications
+    scope = scope.where("id > ?", after) if after.present?
+
+    res = scope.select("min(id), max(id)").limit(1).load[0]
```
Comment (marten) [inline app/models/classifications_export_segment.rb]: Not performant, really:

```
[6] pry(main)> z.set_first_last_classifications(nil)
  Classification Load (24700.2ms)  SELECT  min(id), max(id) FROM "classifications" WHERE "classifications"."workflow_id" = $1 AND "classifications"."completed" = $2 LIMIT 1  [["workflow_id", 338], ["completed", "t"]]
```

I'm open to better solutions if you can think of any?

File: app/operations/classifications_export_segments/create.rb (hunk 2 above)
Comment (marten) [inline app/operations/classifications_export_segments/create.rb]: Don't think so, I'll write a test to make sure.

File: app/controllers/api/v1/classifications_export_segments_controller.rb [hunk 6]
```diff
@@ -0,0 +1,36 @@
+class Api::V1::ClassificationsExportSegmentsController < Api::ApiController
+  require_authentication :all, scopes: [:project]
+
+  before_action :require_access_to_project, only: [:create]
```
Comment (marten) [inline app/controllers/api/v1/classifications_export_segments_controller.rb]: Yeah, index and show are scoped through the `can_through_parent` stuff. But the `create` isn't using the normal stuff.

**Comment by zwolf** [comment]
For currently problematic workflows/projects, we'll want to be able to create smallish segments to get through the backlog of huge annotations/tasks. Is that something we might want to be able to do via rake task?

**Pull request closed without merging**